lib/improver/spotdata/times.py

```diff
@@ -0,0 +1,102 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Plugins written for the Improver site specific process chain."""
+import re
+
+from datetime import datetime as dt
+from datetime import time
+from datetime import timedelta
+
+
+def get_forecast_times(forecast_date=None, forecast_time=None,
+                       forecast_length=None):
+    """
+    Generate a list of python datetime objects specifying the desired forecast
+    times. This list will be created from input specifications if provided.
+    Otherwise defaults are to start today at the most recent 6-hourly interval
+    (00, 06, 12, 18) and to run out to T+144 hours.
+
+    Args:
+    -----
+    forecast_date : string (YYYYMMDD)
+        A string of format YYYYMMDD defining the start date for which forecasts
+        are required. If unset it defaults to today in UTC.
+
+    forecast_time : integer
+        An integer giving the hour on the forecast_date at which to start the
+        forecast output; 24hr clock such that 17 = 17Z for example. If unset it
+        defaults to the latest 6 hour cycle as a start time.
+
+    forecast_length : integer
+        An integer giving the desired length of the forecast output in hours
+        (e.g. 48 for a two day forecast period).
+
+    Returns:
+    --------
+    forecast_times : list of datetime.datetime objects
+        A list of python datetime.datetime objects that represent the times at
+        which diagnostic data should be extracted.
+
+    Raises:
+    -------
+    ValueError : raised if the input date is not in the expected format.
+
+    """
+    date_format = re.compile('[0-9]{8}')
+
+    if forecast_date is not None:
+        if date_format.match(forecast_date) and len(forecast_date) == 8:
+            start_date = dt.strptime(forecast_date, "%Y%m%d").date()
+        else:
+            raise ValueError('Date {} is in unexpected format; should be '
+                             'YYYYMMDD.'.format(forecast_date))
+    else:
+        start_date = dt.utcnow().date()
+
+    if forecast_time is not None:
+        forecast_start_time = dt.combine(start_date, time(forecast_time))
+    else:
+        # If no start hour provided, go back to the nearest multiple of 6
+        # hours (e.g. utcnow = 11Z --> 06Z).
+        forecast_start_time = dt.combine(
+            start_date, time(divmod(dt.utcnow().hour, 6)[0]*6))
+
+    if forecast_length is None:
```

Comment:
I've made it an arg, and set the default in main.py. This removes the need to check the type or enforce an assumed length in here.